packages/compass-generative-ai/src/store/atlas-optin-reducer.ts

```diff
@@ -0,0 +1,325 @@
+import type { Action, AnyAction, Reducer } from 'redux';
+import type { ThunkAction } from 'redux-thunk';
+import { throwIfAborted } from '@mongodb-js/compass-utils';
+import type { AtlasAiService } from '../atlas-ai-service';
+import type { PreferencesAccess } from 'compass-preferences-model';
+
+function isAction<A extends AnyAction>(
+  action: AnyAction,
+  type: A['type']
+): action is A {
+  return action.type === type;
+}
+
+type AttemptState = {
+  id: number;
+  controller: AbortController;
+  promise: Promise<void>;
+  resolve: () => void;
+  reject: (reason?: any) => void;
+};
+
+export type AtlasOptInState = {
+  error: string | null;
+  isModalOpen: boolean;
+  attemptId: number | null;
+  state: 'initial' | 'in-progress' | 'error' | 'canceled' | 'optin-success';
+};
+
+export type GenAIAtlasOptInThunkAction<
+  R,
+  A extends AnyAction = AnyAction
+> = ThunkAction<
+  R,
+  AtlasOptInState,
```

Comment:
The thunk and state here should be the overall reducer state. 
Here's an example in our code base: https://github.com/mongodb-js/compass/blob/2c3f6d068c83866c3bc16c0abd092acdbadd23a6/packages/compass-indexes/src/modules/index.ts#L80
We type the thunk for the entire package's store, with the root state.